proving.go

```diff
@@ -0,0 +1,232 @@
+	var currentPos, subtreeStart position
+	var err error
+	var width uint64
+	value := make([]byte, NodeSize)
+	skipPositions := &positionsStack{}
+	rootHeight := rootHeightFromWidth(readers[0].Width())
+
+	var proof, additionalProof [][]byte
+	for { // Process proven leaves:
+
+		// Get the leaf whose subtree we'll traverse.
+		currentPos, err = provenLeafIndexIt.peek()
+		if err == noMoreItems {
+			// If there are no more leaves to prove - we're done.
+			break
+		}
+
+		// Get the reader for the leaf layer.
+		reader := readers[0]
+
+		// Get indices for the bottom left corner of the subtree and its root, as well as the bottom layer's width.
+		subtreeStart, currentPos, width = subtreeDefinition(currentPos, readers)
+
+		// Prepare reader to read subtree leaves.
+		err = reader.Seek(subtreeStart.index)
+		if err != nil {
+			return nil, errors.New("while preparing to traverse subtree: " + err.Error())
+		}
+
+		// Prepare list of leaves to prove in the subtree.
+		leavesToProve := provenLeafIndexIt.batchPopRelativeIndices(subtreeStart.index, width)
+
+		// Traverse the subtree and append the additional proof nodes to the existing proof.
+		additionalProof, _, err = traverseSubtree(reader, width, hash, leavesToProve)
+		proof = append(proof, additionalProof...)
+		if err != nil {
+			return nil, errors.New("while traversing subtree: " + err.Error())
+		}
+
+		for currentPos.height < rootHeight { // Traverse cache:
+
+			// Check if we're revisiting a node. If we've descended into a subtree and just got back, we shouldn't add
+			// the sibling to the proof and instead move on to the parent.
+			found = skipPositions.PopIfEqual(currentPos)
+			if found {
+				currentPos = currentPos.parent()
+				continue
+			}
+
+			// If the current node sibling is an ancestor of the next proven leaf sibling we should process it's subtree
+			// instead of adding it to the proof. When we reach it again we'll want to skip it.
+			if p, err := provenLeafIndexIt.peek(); err == nil && currentPos.sibling().isAncestorOf(p) {
+				skipPositions.Push(currentPos.sibling())
+				break
+			}
+
+			// Add the current node sibling to the proof:
+
+			// Get the cache reader for the current layer.
+			reader, found = readers[currentPos.height]
+
+			// If the cache wan't found, we calculate the minimal subtree that will get us the required node.
+			if !found {
+				// Find the next cached layer below the current one.
+				for subtreeStart = currentPos.leftChild(); !found; subtreeStart = subtreeStart.leftChild() {
+					reader, found = readers[subtreeStart.height]
+				}
+
+				// Prepare the reader for traversing the subtree.
+				err := reader.Seek(subtreeStart.index)
+				if err != nil {
+					return nil, errors.New("while seeking in cache: " + err.Error() + subtreeStart.String())
+				}
+
+				// Traverse the subtree.
+				width = 1 << (currentPos.height - subtreeStart.height)
+				_, root, err := traverseSubtree(reader, width, hash, nil)
```

Comment:
`root` is the value of `currentPos`, right? If true, it might help to mention it in a comment